workflow/artifacts/common/load_to_stream.go

```diff
@@ -28,7 +28,7 @@ func (w selfDestructingFile) Close() error {
 func LoadToStream(a *wfv1.Artifact, g ArtifactDriver) (io.ReadCloser, error) {
 	log.Infof("Efficient artifact streaming is not supported for type %v: see https://github.com/argoproj/argo-workflows/issues/8489",
 		reflect.TypeOf(g))
-	filename := "/tmp/" + rand.String(32)
+	filename := "/tmp/wfstream-" + rand.String(32)
```

Comment:
Hmm while this source change seems innocuous enough, this does suggest that maybe the test shouldn't be counting files here at all?

Like if the filename should be passed in or returned instead and tested directly, rather than indirectly counting files

Reply:
Perhaps. I considered doing the right thing with [os.CreateTemp](https://pkg.go.dev/os#CreateTemp) or MkdirTemp

But this is a test of a fix to hack around a lack of stream loading in various artifact drivers. The function must conform to the interface there.

The test's primary purpose isn't to check deletion, and mangling the code to make just this portion of the test workable in this case appeared more likely to introduce more bugs than be beneficial. The better choice than doing that would be to remove this test IMHO.

Reply:
> I considered doing the right thing with [os.CreateTemp](https://pkg.go.dev/os#CreateTemp) or MkdirTemp

Yea that would be more correct too



> The function must conform to the interface there.

It actually doesn't seem to need to as far as I can tell. [`OpenStream`](https://github.com/argoproj/argo-workflows/blob/1f0adca09374362d47cf5fd9f19b4b4b2e104d03/workflow/artifacts/common/common.go#L21) is part of our internal interface (which is changeable too), but not `LoadToStream`.

So returning an additional `filename` or similar should be fine, we just may need to change the `OpenStream` calls that use this function to ignore that return arg.



> The test's primary purpose isn't to check deletion, and mangling the code to make just this portion of the test workable in this case appeared more likely to introduce more bugs than be beneficial. The better choice than doing that would be to remove this part of the test IMHO.

IMO I thought changing the filename was actually more intrusive than the return arg proposed above, since that actually changes user-visible outputs (not that a random temp filename should be relied on, but it does cause a change nonetheless), whereas the return arg would be purely internal.

Reply:
A function/method is implying a contract in it's parameters.

Returning the filename from `LoadToStream` would imply that this filename was usable by the caller. It is not, and under no circumstances should be used. I hadn't even considered changing this interface for this reason, and am still happy with this choice.

Reply:
Ah yes, the classic conundrum of "this is only done for testability" 😅.
We do have code that does this in several places: in the UI, exported variables only used in tests, in the Server, `ClientForAuthorization` is passed in purely so it can be mocked in the tests (and plenty more). The latter option would work here too without violating the contract, instead it breaks encapsulation 😕

In general, I agree, we shouldn't change source code for the purpose of testability; that feels backwards. I suppose the problem is that you did do so here already; in functional terms you made the side-effect easier to observe, which the tests do. So the question is "how to correctly test a side-effect?".
I think it's a good safety test, so I'd rather not remove it, although potentially that behavior could be tested differently, e.g. a unit on the `Close` function. Maybe that's the best option?

If we go with your route, I would still prefer to change it to `os.CreateTemp` and make `wfstream-` a constant that the tests use.

Reply:
Sometimes testability improves the code, so then I'm happy to do it.

`os.CreateTemp` returns you a filehandle. `ArtifactDriver.Load()` takes a filename - so can't do that without more invasive surgery.

I could make `wfstream-` a constant, but otherwise this is getting out of hand for a PR to just fix a test flake.

Reply:
> `ArtifactDriver.Load()` takes a filename - so can't do that without more invasive surgery.

oh dang, I missed that, sorry. maybe worth a separate "best practice" change in the future to refactor that, but yea def not worth right now.



> I could make `wfstream-` a constant

I'd be fine with that or remove this part of the test and replacing it with a unit on the `Close` function or something similar that is a more direct test. Let me know what you think. I think either is better than the current

> but otherwise this is getting out of hand for a PR to just fix a test flake.

we tend to discuss more "philosophical" matters of how to write the code properly now and in the future in general. I personally consider that a worthwhile investment, as it is not purely about the flake and may in fact impact how we handle other flakes and other side-effect testing code, i.e. setting precedent and best practices for the future.

Reply:
I've gone with the constant.